skfem/assembly/form/coo_data.py

```diff
@@ -0,0 +1,38 @@
+from dataclasses import dataclass
+from typing import Tuple
+
+from numpy import ndarray
+from scipy.sparse import coo_matrix, csr_matrix
+
+
+@dataclass
+class COOData:
+
+    data: ndarray
+    rows: ndarray
+    cols: ndarray
+    shape: Tuple[int, int]
+
+    @staticmethod
+    def _assemble_scipy_csr(data: ndarray,
+                            rows: ndarray,
+                            cols: ndarray,
+                            shape: Tuple[int, int]) -> csr_matrix:
+
+        K = coo_matrix((data, (rows, cols)), shape=shape)
+        K.eliminate_zeros()
+        return K.tocsr()
+
+    def tocsr(self) -> csr_matrix:
+        """Return a sparse SciPy CSR matrix."""
+        return self._assemble_scipy_csr(
+            self.data,
+            self.rows,
+            self.cols,
+            self.shape,
+        )
+
+    def todense(self) -> ndarray:
```

Comment:
[`coo_matrix.todense`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.coo_matrix.todense.html#scipy.sparse.coo_matrix.todense) actually returns a [`numpy.matrix`](https://numpy.org/devdocs/reference/generated/numpy.matrix.html#numpy.matrix) and 

> It is no longer recommended to use this class, even for linear algebra. Instead use regular arrays. The class may be removed in the future.

It is a subclass of `ndarray` though so probably passes type-tests on that basis.

Maybe go with [`coo_matrix.toarray`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.spmatrix.toarray.html#scipy.sparse.spmatrix.toarray)?